com.unity.netcode.gameobjects/Runtime/Messaging/MessagingSystem.cs

```diff
@@ -150,7 +150,7 @@ internal void HandleIncomingData(ulong clientId, ArraySegment<byte> data, float
                 fixed (byte* nativeData = data.Array)
                 {
                     var batchReader =
-                        new FastBufferReader(nativeData, Allocator.None, data.Count, data.Offset);
+                        new FastBufferReader(nativeData + data.Offset, Allocator.None, data.Count);
```

Comment:
Might have missed something, but why are we moving the `data.Offset` out of the offset argument?

Reply:
Yeah I should have added a comment about that change. Basically the previous behavior was wrong when the `ArraySegment` returned by the transport is not for a segment starting at the beginning of its backing array (`data.Offset` greater than 0).

`FastBufferReader` is written such that the length you pass as the third argument here is the full length of the buffer pointed to by the first argument. But here that's not what we're providing; `data.Count` is the length of the _segment_ of the array we're handling and we were passing a pointer to the backing array of that segment as the buffer.

It might be tempting instead to keep the offset and pass the length of the backing array (`data.Array.Length`) instead of `data.Count`, but that's also wrong since the backing array could go on after the end of the segment. The correct value to pass as the length in this case would be `data.Offset + data.Count` but that's kinda confusing. It's simpler to just make it look to `FastBufferReader` like it's dealing with a clean buffer without any starting offset.

Reply:
Is this a change that should be applied to FastBufferReader itself? When I was writing that part of FastBufferReader I debated with myself on whether the third argument should be the total buffer size or the size to use past the offset. I'm willing to accept I may have chosen wrong here.

Reply:
> This is an immediate point of concern because it means we'll be fixing this for the people who are currently having this issue by introducing new issues to other users. I don't think we should change the meaning of an existing setting in a way that makes existing configurations invalid. IMO, we should deprecate this setting and make a new one with a sensible default so that users who upgrade don't find themselves suddenly broken.

Fair point. Will deprecate that setting and introduce a new one.

Reply:
I don't have any strong opinion on whether that length parameter should be the total length of the buffer or the length past the offset. My concern is more that the documentation of the constructors seems to indicate it's the latter, while internally the code appears to use the former definition.

For the current change, I went with the simplest and smallest targeted fix I could find, considering I'm not familiar with the `FastBufferReader` code and the implications of modifying it.